### PR TITLE
Fix gpt-oss RunAI streamer weight ownership

### PR DESCRIPTION
**Pull request by knlb-tml**
[BEGIN by kunal's robot]

## Motivation

`_load_weights_mxfp4` splits the incoming weights into two lists before loading either of them:

```python
for name, weight in weights:
    if ".experts" in name and ...:
        mxfp4_weights.append((name, weight))
    else:
        normal_weights.append((name, weight))

mxfp4_loaded_params = self._load_mxfp4_experts_weights(mxfp4_weights)
```

Nothing is read until the iterator is exhausted, so every tensor is retained for the length of the stream. As #37368 describes it, the RunAI streamer "may reuse an internal staging buffer between streamed tensors", so a retained view can be overwritten before it is read. The model then loads with no error onto weights that are partly garbage, and every request comes back with non-finite logits.

Same class as the DeepSeek fix (#32913), the GLM5 one (#37370), and the open NVFP4 one (#37371). gpt-oss has neither a guard nor an issue.

## Modifications

- yield the expert weights through a generator instead of collecting them. They are the bulk of the checkpoint and `_load_mxfp4_experts_weights` copies each into its parameter as it arrives, so they never outlive the staging buffer
- copy the non-expert weights, which are still retained after the stream ends, to the host when they carry `RUNAI_STREAMER_TENSOR_ATTR`. Distributed streaming yields device tensors for object-storage URIs, so cloning them in place would add their own GiB to peak GPU usage, outside the streamer's configured limit
- add `test/registered/unit/models/test_gpt_oss_runai_ownership.py`, two hermetic tests: expert weights are consumed as they are yielded, and a retained streamed weight is copied out of the buffer while an unmarked one is passed through untouched. Both fail on the unpatched loader, the first with `[3, 3, 3] != [1, 2, 3]`, the second with `unexpectedly identical`

## Accuracy Tests

`openai/gpt-oss-20b`, single GPU, `tp_size=1`, offline `sgl.Engine`, greedy, the same local snapshot in every arm:

| load format | `RUNAI_STREAMER_MEMORY_LIMIT` | output for "The capital of France is" |
|---|---|---|
| `runai_streamer` | default | ` Paris."\n\nSure! Here's a simple example…` |
| `runai_streamer` | 1158266880 | `!!!!!!!!!!!!!!!!!!!!!!!!` |
| `safetensors` | default | ` Paris."\n\nSure! Here's a simple example…` |
| `runai_streamer` + this patch | 1158266880 | ` Paris."\n\nSure! Here's a simple example…` |

`!!!!` is greedy sampling on non-finite logits. 1158266880 is the smallest limit the streamer accepts for this checkpoint, since it refuses anything below its largest chunk; at that size the buffer is recycled many times over. At the default limit a 12 GiB checkpoint never wraps it, which is why this hides at small scale. It surfaced first on `gpt-oss-120b`, where the default buffer wraps on its own and every request was non-finite.

I have not run a GSM8K evaluation. Happy to add one if that is wanted before merge.

## Speed Tests and Profiling

No speed impact expected: the same bytes are read in the same order, and the only added work is cloning the non-expert weights, which are a small fraction of the checkpoint. No benchmark run.

## Checklist

- [x] Format your code according to the Format code with pre-commit. `pre-commit run --files` on both changed files: isort, ruff, ruff format, codespell and the registered-test validators all pass.
- [x] Add unit tests according to the Run and add unit tests. Added under `test/registered/unit/models/`, registered with `register_cpu_ci`, verified to fail without the fix.
- [ ] Update documentation according to Write documentations. No documented behaviour changes.
- [x] Provide accuracy and speed benchmark results. Accuracy above; speed unaffected and not benchmarked.
- [x] Follow the SGLang code style guidance.

## Notes

`_canonicalize_weights` still opens with `dict(weights_in)`, so the non-mxfp4 path has the same exposure. I left it alone because everything in that function is retained by design, so the equivalent fix clones the whole checkpoint into host memory, which is a real regression rather than a few GiB. Happy to follow up separately, or to defer to `stream_files(..., owned=True)` if that lands.

If #37371 merges first, the local `_clone_if_runai_streamed_tensor` here should move to the shared `sglang.srt.model_loader.runai_utils` helper it introduces rather than becoming a fourth copy.

[END by kunal's robot]

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34618907481](https://github.com/sgl-project/sglang/actions/runs/34618907481)
Latest PR Test (Extra): :white_check_mark: [Run #34618907058](https://github.com/sgl-project/sglang/actions/runs/34618907058)
Latest PR Test (AMD ROCm 10): :x: [Run #34618907267](https://github.com/sgl-project/sglang/actions/runs/34618907267)

### CODE REVIEW

File: python/sglang/srt/models/gpt_oss.py [hunk 1]
```diff
@@ -953,20 +956,27 @@ def load_weights(
...
+            for name, weight in weights:
+                if (
+                    ".experts" in name
+                    and self.quant_config is not None
+                    and self.quant_config.get_name() == "mxfp4"
+                ):
+                    yield name, weight
+                else:
+                    normal_weights.append(
+                        (name, _clone_if_runai_streamed_tensor(weight))
```
[the remark this answers is not in the document]
Comment (knlb-tml) [reply python/sglang/srt/models/gpt_oss.py]: [BEGIN by kunal's robot]

Good catch, fixed in 22e35dc. `_is_distributed` defaults to `is_object_storage_path`, so exactly the loads this patch targets are the ones yielding device tensors, and cloning in place would have traded a correctness bug for an OOM.

The retained weights are now copied to the host rather than cloned on device. They only need to be somewhere the streamer cannot overwrite, and they are held until the stream ends, so device memory is the wrong place for them. Peak GPU usage is unchanged from before this PR; the cost is a few GiB of host memory and one D2H copy of the non-expert weights.

The unit test now also asserts the retained copy lands on CPU.

[END by kunal's robot]

**Comment by ekzhang** [comment]
@knlb-tml did your robot make a stray PR?

**Comment by knlb-tml** [comment]
Nope, sam asked me to make this upstream explicitly -- just cleaning it up based on chatgpt feedback

**Review by ispobock — APPROVED** [review]

**Comment by knlb-tml** [comment]
ty!
